### PR TITLE
Added currency conversions and more skinning

### PR DESCRIPTION
**Pull request by einsteinx2**

### CODE REVIEW

File: Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    }
+    
+    func convert(amount: Double, from: Currency, to: Currency, source: ExchangeRateSource, iteration: Int = 0) -> Double? {
+        guard iteration < 5 else {
+            return nil
+        }
+        
+        log.debug("converting from \(from) to \(to) source \(source)")
+        
+        if let exchangeRates = exchangeRates(forSource: source) {
```
Comment (wolffan) [inline Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift]: that's the duplicated logic we want on the same lib right?

File: Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                    return convert(amount: amount, from: from, to: to, source: .coinbaseGdax, iteration: iteration + 1)
+                }
+            }
+        }
+        
+        return nil
+    }
+    
+    func updateExchangeRates() {
+        let url = URL(string: "https://balance-server.appspot.com/exchangeRates")!
```
Comment (wolffan) [inline Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift]: maybe this can be put out into a constant (beginning  of file)

File: Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        let task = certValidatedSession.dataTask(with: url) { maybeData, maybeResponse, maybeError in
+            // Make sure there's data
+            guard let data = maybeData, maybeError == nil else {
+                log.error("Error updating exchange rates, either no data or error: \(String(describing: maybeError))")
```
Comment (wolffan) [inline Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift]: this should be promoted? if ther's no data it's likely its our error on backend

File: Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                return
+            }
+            
+            // Parse and cache the data
+            if self.parse(data: data) {
+                self.persist(data: data)
+                NotificationCenter.postOnMainThread(name: Notifications.exchangeRatesUpdated)
```
Comment (wolffan) [inline Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift]: not a big fan of this... but i haven't looked at this part architecture

File: Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            }
+        }
+        task.resume()
+    }
+    
+    @discardableResult func parse(data: Data) -> Bool {
+        // Try to parse the JSON
+        guard let tryExchangeRatesJson = try? JSONSerialization.jsonObject(with: data) as? [String: Any], let exchangeRatesJson = tryExchangeRatesJson, exchangeRatesJson["code"] as? Int == BalanceError.success.rawValue else {
+            log.error("Error parsing exchange rates, failed to parse json data")
```
Comment (wolffan) [inline Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift]: again this may be related to our server, shouldn't we use HockeyApp to log those so we can be sure there's no problem in our end?

File: Balance/Shared/Data Model/Exchange Rates/ExchangeRate.swift [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+//        try container.encode(rate, forKey: .rate)
+//    }
+}
+
+public extension Array where Element == ExchangeRate {
+    public func contains(from: Currency, to: Currency) -> Bool {
+        return self.contains(where: {$0.from == from && $0.to == to})
+    }
+    
+    public func rate(from: Currency, to: Currency) -> Double? {
```
Comment (wolffan) [inline Balance/Shared/Data Model/Exchange Rates/ExchangeRate.swift]: shouldn't this be non optional? and then we call contains and return a non optional?
Otherwise contains doesn't make much sense no?

From outisde you call contains and then the rate... my mind

File: Balance/Shared/Data Model/Exchange Rates/ExchangeRate.swift [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+import Foundation
+
+public struct ExchangeRate { //: Codable {
+    let source: ExchangeRateSource
+    let from: Currency
+    let to: Currency
+    let rate: Double
+    
+//    enum CodingKeys: String, CodingKey {
```
Comment (wolffan) [inline Balance/Shared/Data Model/Exchange Rates/ExchangeRate.swift]: is this WIP?

Also the exchangeRate struct was codable but is no more?

File: Balance/Shared/Data Model/Exchange Rates/ExchangeRateSource.swift [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    case bitfinex     = 3
+    case kraken       = 4
+    
+    // Fiat
+    case fixer        = 10001
+    
+    public var url: URL {
+        switch self {
+        case .coinbaseGdax:
+            return URL(string: "https://api.coinbase.com/v2/prices/usd/spot")!
```
Comment (wolffan) [inline Balance/Shared/Data Model/Exchange Rates/ExchangeRateSource.swift]: this shouldn't be here? should it?
That is just backend logic...

File: Balance/Shared/Data Model/Other Models/SimpleCache.swift [hunk 9]
```diff
@@ -0,0 +1,43 @@
+//
+//  SimpleCache.swift
+//  BalanceServer
+//
+//  Created by Benjamin Baron on 10/15/17.
+//
+
+import Foundation
+
+public class SimpleCache<KeyType: Hashable, ValueType: Any> {
```
Comment (wolffan) [inline Balance/Shared/Data Model/Other Models/SimpleCache.swift]: again why did we take this from the backend? The cache should be on client... NSURLrequest I think does cache as well for http

File: Balance/Shared/Data Model/Singleton/Defaults.swift [hunk 10]
```diff
@@ -291,6 +291,24 @@ class Defaults {
...
     }
     
+    var masterCurrency: Currency {
+        get {
+            if let raw = defaults.object(forKey: Keys.masterCurrency) as? String {
+                return Currency.rawValue(raw)
+            }
+            
+            // Default to current local if nothing is set
+            if let currencyCode = NSLocale.current.currencyCode {
```
Comment (wolffan) [inline Balance/Shared/Data Model/Singleton/Defaults.swift]: 👍

File: Balance/Shared/Data Model/Singleton/ServerMessage.swift [hunk 11]
```diff
@@ -8,7 +8,7 @@
 
 import Foundation
 
-private let messageUrl = URL(string: "https://balancemy.money/institutions/message.json")!
+private let messageUrl = URL(string: "https://balance.nyc3.digitaloceanspaces.com/message/message.json")!
```
Comment (wolffan) [inline Balance/Shared/Data Model/Singleton/ServerMessage.swift]: why the change?

I get this:
{
	"id": 4,
	"expiration": 0,
	"title": "Title Goes Here",
	"message": "Message goes here.\n\nYou can use backslash n to make a new line.",
	"okButtonTitle": "OK"
}

File: Balance/Shared/Data Model/Singleton/Utils.swift [hunk 12]
```diff
@@ -202,18 +202,18 @@ fileprivate var amountFormatter: NumberFormatter = {
 }()
 
 // TODO: Add proper locale support for currency symbol location
-func amountToString(amount: Int, currency: Currency, showNegative: Bool = false)  -> String {
+func amountToString(amount: Int, currency: Currency, showNegative: Bool = false, showCodeAfterValue: Bool = false)  -> String {
```
Comment (wolffan) [inline Balance/Shared/Data Model/Singleton/Utils.swift]: we didn't have tests for this? :(

File: Balance/Shared/Data Model/Sqlite Models/Repositories/TransactionRepository.swift [hunk 13]
```diff
@@ -270,72 +261,62 @@ struct TransactionRepository: ItemRepository {
...
+                if transactions.count > 0 {
+                    transactionsByDate[previousDate] = transactions
+                    counts.append(transactions.count)
+                }
+            } catch {
+                log.severe("Error getting transactions by date: " + db.lastErrorMessage())
             }
-            
-            addTransactions(pending: true)
-            addTransactions(pending: false)
```
Comment (wolffan) [inline Balance/Shared/Data Model/Sqlite Models/Repositories/TransactionRepository.swift]: all the pending stuff was related to Plaid?

File: Balance/Shared/Data Model/Syncing/Syncer.swift [hunk 14]
```diff
@@ -61,6 +61,7 @@ class Syncer {
         var syncingInstitutions = institutions
         
         if !canceled, let institution = syncingInstitutions.popLast() {
+            print("institutions: \(institutions) syncingInstitutions: \(syncingInstitutions)")
```
Comment (wolffan) [inline Balance/Shared/Data Model/Syncing/Syncer.swift]: again, log? :)

File: Balance/macOS/View Controllers/Tabs/View Controllers/TabsViewController.swift [hunk 15]
```diff
@@ -291,14 +236,14 @@ class TabsViewController: NSViewController {
...
-            } else {
-                tabButton.deactivate()
-            }
-        }
+//        for i in 0...tabButtons.count-1 {
+//            let tabButton = tabButtons[i]
+//            if i == tabIndex {
+//                tabButton.activate()
+//            } else {
+//                tabButton.deactivate()
```
Comment (wolffan) [inline Balance/macOS/View Controllers/Tabs/View Controllers/TabsViewController.swift]: delete?

File: Balance/macOS/View Controllers/Tabs/View Controllers/TabsViewController.swift [hunk 16]
```diff
@@ -410,11 +363,12 @@ class TabsViewController: NSViewController {
...
                     if let characters = event.charactersIgnoringModifiers {
                         if event.modifierFlags.contains(NSEvent.ModifierFlags.command) && characters.length == 1 {
-                            if let intValue = Int(characters), intValue > 0 && intValue <= self.tabButtons.count {
-                                // Select tab
-                                self.tabAction(self.tabButtons[intValue - 1].button)
-                                return nil
-                            } else if characters == "," {
+//                            if let intValue = Int(characters), intValue > 0 && intValue <= self.tabButtons.count {
+//                                // Select tab
+//                                self.tabAction(self.tabButtons[intValue - 1].button)
```
Comment (wolffan) [inline Balance/macOS/View Controllers/Tabs/View Controllers/TabsViewController.swift]: deleet?

File: BalanceUnitTests/Exchange Rates/CurrentExchangeRatesTests.swift [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+        let data = TestHelpers.loadData(filename: "ExchangeRates.json")
+        currentExchangeRates.parse(data: data)
+        
+        //when
+        let ethUsd = currentExchangeRates.convert(amount: 1.0, from: .eth, to: .usd, source: .poloniex)
+        let btcUsd = currentExchangeRates.convert(amount: 1.0, from: .btc, to: .usd, source: .poloniex)
+        let ltcUsd = currentExchangeRates.convert(amount: 1.0, from: .ltc, to: .usd, source: .poloniex)
+        
+        //then
+        XCTAssertNotNil(ethUsd)
```
Comment (wolffan) [inline BalanceUnitTests/Exchange Rates/CurrentExchangeRatesTests.swift]: Tests <3

File: Balance/Shared/Data Model/APIs/Coinbase/CoinbaseApi.swift [hunk 18]
```diff
@@ -296,7 +296,7 @@ extension Institution {
                 refreshToken = dictionary["refreshToken"] as? String
             }
             
-            print("get refreshTokenKey: \(refreshTokenKey)  refreshToken: \(String(describing: refreshToken))")
+            //print("get refreshTokenKey: \(refreshTokenKey)  refreshToken: \(String(describing: refreshToken))")
```
[the remark this answers is not in the document]
Comment (einsteinx2) [reply Balance/Shared/Data Model/APIs/Coinbase/CoinbaseApi.swift]: For these ones, I had just commented them out so that I could copy my Xcode debug log without giving away a bunch of info. These should be able to stay as log.debug statements.

[the remark this answers is not in the document]
Comment (einsteinx2) [reply Balance/Shared/Data Model/APIs/Coinbase/CoinbaseApi.swift]: Pushed a commit that replaces all print statements by either removing them, log.debug for xcode builds only, and log.info/error for all builds

File: Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift (hunk 1 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift]: Yep exactly. As discussed I'll back-port to the server side and work from there, then we'll move all this to a shared framework soon.

File: Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift (hunk 2 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift]: Agreed

File: Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift (hunk 3 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift]: In my mind, no because if the backend is temporarily down, we'll just use the latest cached exchange rate data rather than specifically reporting that exchange rates are not updating especially since there's nothing the user can do about it (unlike if an account fails to sync which may need a new api key or something). 

In any case, the backend should never be down for any length of time, and if it does go down, various other app functions like keystore syncing and profile sign in, etc, will also not work, so we'll have way bigger problems than out of date exchange rates.

File: Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift (hunk 4 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift]: What don't you like about it?

File: Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift (hunk 5 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Exchange Rates/CurrentExchangeRates.swift]: I'll add a hockeyapp event to track these. They should never happen unless our backend breaks, or if the cached data file gets corrupted on the user's machine.

File: Balance/Shared/Data Model/Exchange Rates/ExchangeRate.swift (hunk 7 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Exchange Rates/ExchangeRate.swift]: Actually it was a dead end, I was coing to make it Codable for caching but realized I could save the json response much more simply. I'll remove it.

File: Balance/Shared/Data Model/Exchange Rates/ExchangeRate.swift (hunk 6 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Exchange Rates/ExchangeRate.swift]: I only added the contains function as a convenience function for when I really only want to know if it contains something, and in fact I had only added it to use in one particular bit of code that doesn't even exist anymore, so it's not used at all.

The rate function is the main function and must be optional because it doesn't always have an answer. The idiomatic swift way of using it would be `if let rate = exchangeRate.rate(from...) {` and then do something if it has a rate, which both tells you if it contains it and gives it to you if it does in one operation.

File: Balance/Shared/Data Model/Exchange Rates/ExchangeRateSource.swift (hunk 8 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Exchange Rates/ExchangeRateSource.swift]: Yep I just copied over the files. This will all be in the shared framework soon and if we want we can add this var as an extension inside the server code instead of including it in the shared code.

File: Balance/Shared/Data Model/Other Models/SimpleCache.swift (hunk 9 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Other Models/SimpleCache.swift]: We cache on both the client and server. On the server, we cache the json response to the client so that we only ever hit the db one per minute even if a million users are trying to update their rates constantly. We can't use NSURLRequest for that since it's caching response data not request data.

On the client side, I wanted finer grained control over caching than what NSURLRequest would provide and I want to cache the parsed data not have to re-parse it for every convert call, so I use simplecache there to save the data. We could use this to cache any kind of data anywhere we want.

File: Balance/Shared/Data Model/Singleton/ServerMessage.swift (hunk 11 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Singleton/ServerMessage.swift]: In general, any file that is not really related to our website, I've moved to Dropbox's Spaces (their AWS S3 competitor). That way if we want to update the Sparkle xml feed or this message file, we can do it without pushing to Gatsby/Netlify.

The file as it is now is a placeholder, the expiration is way in the past (timestamp 0), so it never gets displayed. If we ever need to display an emergency message like we did with the legacy app, we can use this mechanism.

File: Balance/Shared/Data Model/Singleton/Utils.swift (hunk 12 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Singleton/Utils.swift]: Nope, this is modified from old legacy code, but we should definitely get test coverage for this as we continue to add more tests. Luckily, it's pretty obvious when it doesn't work as it would break all kinds of UI.

File: Balance/Shared/Data Model/Sqlite Models/Repositories/TransactionRepository.swift (hunk 13 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Sqlite Models/Repositories/TransactionRepository.swift]: Yeah, right now we don't get a pending flag from our data sources. Also in general, we decided on using a simpler transactions table and then store metadata in a separate table. Right now we don't have pending metadata to store at all, so I removed it.

File: Balance/Shared/Data Model/Syncing/Syncer.swift (hunk 14 above)
Comment (einsteinx2) [reply Balance/Shared/Data Model/Syncing/Syncer.swift]: fixed

File: Balance/macOS/View Controllers/Tabs/View Controllers/TabsViewController.swift (hunk 15 above)
Comment (einsteinx2) [reply Balance/macOS/View Controllers/Tabs/View Controllers/TabsViewController.swift]: Yep, removed

File: Balance/macOS/View Controllers/Tabs/View Controllers/TabsViewController.swift (hunk 16 above)
Comment (einsteinx2) [reply Balance/macOS/View Controllers/Tabs/View Controllers/TabsViewController.swift]: Actually meant to uncomment this and fix it. Fixed.

File: BalanceUnitTests/Exchange Rates/CurrentExchangeRatesTests.swift (hunk 17 above)
Comment (einsteinx2) [reply BalanceUnitTests/Exchange Rates/CurrentExchangeRatesTests.swift]: Yep! :)

**Comment by einsteinx2** [comment]
Pushed a bunch of changes responding to all of your comments. Should be good to merge now.
